Improve performance of font fetching

pvollan opened the pull request:
#### c58276b87f4dcd868107f67931d1fdb50db4d27b
<pre>
Improve performance of font fetching
<a href="https://bugs.webkit.org/show_bug.cgi?id=294405">https://bugs.webkit.org/show_bug.cgi?id=294405</a>
<a href="https://rdar.apple.com/153920635">rdar://153920635</a>

Reviewed by Chris Dumez.

In &lt;<a href="https://commits.webkit.org/296142@main">https://commits.webkit.org/296142@main</a>&gt;, we improved the performance of font registration in the WebContent
process, by only registering fonts on demand. This patch is further improving the performance, but this time in
the UI process. Instead of creating the mapping between font names and URLs in-process, this patch instead gets
this information from a system service, where this data already is prepared and stored in memory. Local
measurements show that this is approximately 10x faster than doing it in-process.

This patch also fixes an issue where we would not register all font files for a given font family. This is
addressed by creating a mapping between font family names and font names, which is sent to the WebContent
process.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/cocoa/SoftLinking.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::WTF_REQUIRES_LOCK):
(WebCore::FontCache::createFontPlatformData):
(WebCore::userInstalledFontFamilyMap):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAdditionalFonts):
(SOFT_LINK_CONSTANT_MAY_FAIL): Deleted.
(WebKit::addUserInstalledFontURLs): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerFontMap):

Canonical link: <a href="https://commits.webkit.org/296738@main">https://commits.webkit.org/296738@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1a3952124e035f4292104831047f2b4e4dbe262

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29099 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114646 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59671 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29778 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37687 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83180 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112389 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98574 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63638 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16717 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59265 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101940 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93084 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16759 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117760 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107997 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36482 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27011 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92191 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36853 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94836 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92007 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36939 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14685 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32287 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17662 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36375 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41851 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132263 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36046 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35828 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39385 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1486,79 +1493,68 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+        if (fontNames != fontFamilyMap.end())
+            fontNames->value.append(fontNameLowerCase);
+        else {
+            Vector<String> fontNames;
```
`Vector<String> fontNames { fontNameLowerCase };` I think would be and would be more efficient then appending on the next line.

pvollan replied:
Ah, good point, fixed in latest patch.

Thanks for reviewing!

pvollan committed:
Improve performance of font fetching
https://bugs.webkit.org/show_bug.cgi?id=294405
rdar://153920635

Reviewed by Chris Dumez.

In <https://commits.webkit.org/296142@main>, we improved the performance of font registration in the WebContent
process, by only registering fonts on demand. This patch is further improving the performance, but this time in
the UI process. Instead of creating the mapping between font names and URLs in-process, this patch instead gets
this information from a system service, where this data already is prepared and stored in memory. Local
measurements show that this is approximately 10x faster than doing it in-process.

This patch also fixes an issue where we would not register all font files for a given font family. This is
addressed by creating a mapping between font family names and font names, which is sent to the WebContent
process.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/cocoa/SoftLinking.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::WTF_REQUIRES_LOCK):
(WebCore::FontCache::createFontPlatformData):
(WebCore::userInstalledFontFamilyMap):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAdditionalFonts):
(SOFT_LINK_CONSTANT_MAY_FAIL): Deleted.
(WebKit::addUserInstalledFontURLs): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerFontMap):

Canonical link: https://commits.webkit.org/296738@main
